### DISCUSSION TITLE
ClientServerRequestStreamExampleTcp / requestresponse questions

### DISCUSSION THREAD

**Discussion by schoko11** [rsocket/rsocket-js#270] [Q&A] [answered]
Hi @viglucci 
thanks for this project!

I try to setup a browser based app to control a hardware (kemper amplifier, where i wiresharked the ip traffic from / to the android app).
So it is just a tcp connection with quit a lot of data, the data is not encrypted. 
The hardware is connected to the local network with a repeater, ip and port from wireshark.

I tried the above examples(which are not not browser compatible just yet i think, but anyways ) with only the client part called.

1.) In wireshark i can see that the hardware is sending a tcp packet with data, how can i subscribe to that data (when parsing the data i need to do further actions).
2.) when using one of both examples i can see two messages 
   2.1.) while this one does not matter, but i wonder how avoid this ?
```
0000   02 0c 43 01 08 72 68 54    5a d0 d2 25 08 00 45 00   ..C..rhTZ..%..E.
0010   00 6f 86 70 40 00 40 06    32 69 c0 a8 00 52 c0 a8   .o.p@.@.2i...R..
0020   00 0d df a2 16 5f 96 cc    0a 77 00 00 6c 0c 50 18   ....._...w..l.P.
0030   fa f0 82 11 00 00 00 00    44 00 00 00 00 04 00 00   ........D.......
0040   01 00 00 00 00 ea 60 00    04 93 e0 18 61 70 70 6c   ......`.....appl
0050   69 63 61 74 69 6f 6e 2f    6f 63 74 65 74 2d 73 74   ication/octet-st
0060   72 65 61 6d 18 61 70 70    6c 69 63 61 74 69 6f 6e   ream.application
0070   2f 6f 63 74 65 74 2d 73    74 72 65 61 6d            /octet-stream
```
2.2.) here in my sent message there are prepend hex before "hello World" -> how can this be avoided (this ? https://www.wireshark.org/lists/wireshark-users/200805/msg00206.html) ? 
```
0000   02 0c 43 01 08 72 68 54    5a d0 d2 25 08 00 45 00   ..C..rhT  Z..%..E.
0010   00 3c 86 71 40 00 40 06    32 9b c0 a8 00 52 c0 a8   .<.q@.@.  2....R..
0020   00 0d df a2 16 5f 96 cc    0a be 00 00 6c af 50 18   ....._..  ....l.P.
0030   fa 4d 81 de 00 00 00 00    11 00 00 00 01 10 00 48   .M......  .......H
0040   65 6c 6c 6f 20 57 6f 72    6c 64                     ello Wor  ld
```
`00 00    11 00 00 00 01 10 00`

3.) my first tryouts were made with node's net module, and also as with the net module i have "tcp retransmission errors"  when sending data to the hardware, so assuming i can only receive but cannot send tcp messages. Any Idea why (i can ping the hardware) ?

As of point 3.) because it is the most important point for me, the communication between hardware and android app is:

- hardware sends udp
- android sends udp back 
```
0000   ff ff ff ff ff ff b6 69 fe a0 c2 1e 08 00 45 00   .......i......E.
0010   00 3e 63 e1 40 00 40 11 54 72 c0 a8 00 0c c0 a8   .>c.@.@.Tr......
0020   00 ff eb f1 16 5f 00 2a 17 0d 44 53 43 56 16 4d   ....._.*..DSCV.M
0030   41 43 23 30 30 3a 30 30 3a 30 30 3a 30 30 3a 30   AC#00:00:00:00:0
0040   30 3a 30 30 07 50 4f 4c 4c 3a 29 00               0:00.POLL:).
```

- syn - syn,ack - ack
- hardware sends tcp and asks for type
- android sends type of client starts.

**maybe the hardware needs a udp message to accept tcp conversation ? can i send a udp message with rsocket too ?**

Thank you (of course i could share some example when it works, in case you could need it)!

**Comment by viglucci** [comment, marked as answer]
Hiya @schoko11 

Happy to hear you are looking at rsocket-js for your project. After reading your question(s), I am a tad bit uncertain though where in your solution you are attempting to leverage RSocket/rsocket-js.

Unless I am mistaken, it sounds like your Kemper device is transmitting data using TCP, but does not sounds like it utilizes the RSocket protocol. If you want to use RSocket to communicate with the device, or as a portion of the communications you'll need to convert the TCP streamss from the device into RSocket compatible streams. Given that, you'll need to find some way to understand how the protocol from the Kemper device functions, and then from there you may be able to write a layer/service to do the translation, however trying to connect a RSocket client directly to the device is unlikely to be successful.

Are you trying to use RSocket for this project, or mainly just asking general questions about TCP with nodejs/javascript?

If you are looking to leverage RSocket/rsocket-js for this project, based on my limted understanding of what it sounds like you are looking to accomplish, this is how I would suspect you might approach it:

![image](https://github.com/rsocket/rsocket-js/assets/6305490/7570b540-83b3-4fcb-9173-0b450024409f)

**A few additional things I can address from your questions above:**

> I tried the above examples(which are not not browser compatible just yet i think, but anyways ) with only the client part called.

IF you decide you want to use RSocket as part of your project, and that usage is to occur within a browser, then you'll need to use the WebSocket transport rather than the TCP transport. TCP is not supported natively in browser environments.

That said, you again aren't going to be able to connect the RSocket WebSocket client directly to your kemper device, you are going to need to author a which exposes a RSocket WebSocket server interface to satisfy the needs of the RSocket WebSocket client.

> 2.2.) here in my sent message there are prepend hex before "hello World" -> how can this be avoided (this ?

The RSocket protocol requires certain framing on packet types. This is part of the protocol and not something that can be avoided if using RSocket-js.

> 1.) In wireshark i can see that the hardware is sending a tcp packet with data, how can i subscribe to that data (when parsing the data i need to do further actions).

The net module from Node is the correct way to create a TCP client (or server), however, again you won't be able to run this within a browser environment.

https://nodejs.org/api/net.html#netcreateconnection:~:text=to%20this%20function.-,net.createConnection(options%5B%2C%20connectListener%5D),-%23

> 3.) my first tryouts were made with node's net module, and also as with the net module i have "tcp retransmission errors" when sending data to the hardware, so assuming i can only receive but cannot send tcp messages. Any Idea why (i can ping the hardware) ?

Honestly I'm not sure and I don't think I'll be able to assist you much with troubleshooting these particular issues. I can't claim to be deeply familiar with TCP or the net module, apart from leveraging them to the extent that was needed to implement the RSocket TCP Client/Server transports for this package.

> maybe the hardware needs a udp message to accept tcp conversation ? can i send a udp message with rsocket too ?

RSocket supports UDP at the protocol level, however there is not a UDP transport currently implemented for rsocket-js. Regardless, even if there were it would not be suitable for this usecase unless the RSocket protocol was understood on the receiving end.

**Reply by schoko11** [reply]
Hi @viglucci thanks for the detailed reply.

i was not aware of the rsocket protocoll, and addying another layer (however that would look like) seems not a good idea(and adds complexity).

You should be able to run node apps in the browser with browserify (i am not sure if this could be bundled with esbuild though).

 
We had a partly similar project at work where my colleague could implement a debugger by research, sending and receiving on one port.
I had a talk to him today, and i also took a look at the code.... So i assuming i missed the server part on the net module. (or the missing the udp send, which i can mimic with netcat).
```
//receive
    this.server = net.createServer();
        this.server.listen(this.debugPort).on('connection', socket => {
            this.socket = socket;
                  socket.on('data', buffer => {
...
//send
public sendPacket(socket: net.Socket) {

   socket.write(this.packet);
```


> however trying to connect a RSocket client directly to the device is unlikely to be successful.

I am not sure what you mean ?

> Are you trying to use RSocket for this project, or mainly just asking general questions about TCP with nodejs/javascript?

Seems like both cases... :-)

Thank you for your answers...

**Reply by viglucci** [reply]
> thanks for the detailed reply.

No problem! 😄 

> > however trying to connect a RSocket client directly to the device is unlikely to be successful.
> 
> I am not sure what you mean ?

Sorry for the confusion, I was referencing how rsocket-js is intended to be used only with servers implementing the RSocket protocol, which it doesn't sound like the Kemper device implements.

**Reply by viglucci** [reply]
> You should be able to run node apps in the browser with browserify (i am not sure if this could be bundled with esbuild though).

The caveat here is that some APIs for node can be satisfied with a polyfill, which browserify would do for you, however the net module is sadly not one of them.

https://stackoverflow.com/a/30831270/2165116

I would be surprised to hear that your co-worker got this working. Maybe they were using another approach such as WebRTC or WebSocket?

**Comment by schoko11** [comment]
> > You should be able to run node apps in the browser with browserify (i am not sure if this could be bundled with esbuild though).
> 
> The caveat here is that some APIs for node can be satisfied with a polyfill, which browserify would do for you, however the net module is sadly not one of them.
> 
> https://stackoverflow.com/a/30831270/2165116
> 
> I would be surprised to hear that your co-worker got this working. Maybe they were using another approach such as WebRTC or WebSocket?

Well it is "sockets" with nodes net module (have to dig into this topics too), some pieces underneath
```
  this.server = net.createServer();
        this.server.listen(this.debugPort).on('connection', socket => {
            this.socket = socket;
                  socket.on('data', buffer => {
...

   this.server.on('error', () => {
            this.shutdown();
        });

   this.server.on('close', () => {
         this.shutdown();
        });
```

i now have a basic thing tested with 
```
let client = new net.socket();
client.connect({port .. host})

client.on("data, data => {
...
client.write("xxx")
```

next up would be a tryout of this with socket.io or alternatives...
